proposals/1412/client_proxy.md

```diff
@@ -0,0 +1,157 @@
+# Problem Statements
+Currently, OpenLineage integrations are directly sending events to the configured backend, which is not flexible enough for some situations such as
+- Not responsive backend due to an outage or a network issue
+- Long latency of the API that adds extra delay to the processing
+
+The first scenario will likely cause the data loss which we would like to avoid as much as possible. The second one is less critical and could be mitigated by implementing asynchronous integrations, but it is very hard to do so for every single integrations.
+
+# Proposal
+In this proposal, I suggest implementing OpenLineage Proxy that is lightweight, fast, and can be run as a sidecar along with the processing engine. This sidecar proxy will be run in the same network as the processing engine and the network delay will be minimal. Integrations may or may not implement asynchronous call, but anyway it is simple super fast call which brings very minimal extra latency. The sidecar proxy is also responsible for dealing with the backend outage. Even if the backend is not able to respond, the sidecar will keep the OpenLineage events locally and retry them. This prevents the data loss even in the case of the backend outage, and we are able to have a full picture of the lineage.
+
+# Requirements
```

Comment:
We might want to also call out that proxy client must be highly configurable (buffer flushing, `N` retries, errors, etc)?

Reply:
This section is already mentioning about the buffer limit configuration, so I think it implies that it is configurable. For the retry, I think the proxy should indefinitely retry until the backend is available as long as the failure was retryable, so making retry count configurable doesn't make sense. I will add a `SHOULD` statement for the buffer flushing.